### PR TITLE
MobManager

### PR DESCRIPTION
**Pull request by Sefiraat**
This PR is to try to provide a way to spawn more mobs faster but correctly manage the amount and removal over time/shutdown

Todo:
- [x] Remove (manager) mobs when named
- [x] Remove (manager) mobs that die or are removed via other means
- [x] Remove (manager and world) that unload
- [x] Remove (world) on addon disable
- [x] Set up a categorisation system to enable a good distribution of mob types

### CODE REVIEW

File: src/main/java/dev/sefiraat/netheopoiesis/managers/MobManager.java [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    public MobManager() {
+        Preconditions.checkArgument(instance == null, "Cannot create a new instance of the MobManager");
+        instance = this;
+    }
+
+    @Nonnull
+    @Unmodifiable
+    public static List<UUID> getPlayerMobs(@Nonnull Player player) {
+        return getPlayerMobs(player.getUniqueId());
+    }
```
Comment (J3fftw1) [inline src/main/java/dev/sefiraat/netheopoiesis/managers/MobManager.java:38]: This is a weird name
do we maybe want to make our own #getPlayer and do getPlayer.getMobs?

**Review by J3fftw1 — CHANGES_REQUESTED** [review]

File: src/main/java/dev/sefiraat/netheopoiesis/managers/MobManager.java (hunk 1 above)
Comment (Sefiraat) [reply src/main/java/dev/sefiraat/netheopoiesis/managers/MobManager.java:38]: #getMobsByPlayer() ?

Comment (J3fftw1) [reply src/main/java/dev/sefiraat/netheopoiesis/managers/MobManager.java:38]: yea sounds good aswell

File: src/main/java/dev/sefiraat/netheopoiesis/listeners/ManagedMobListener.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+            MobManager.getInstance().removeMob(livingEntity, false);
+        }
+    }
+
+    @EventHandler
+    public void onChunkUnload(@Nonnull ChunkUnloadEvent event) {
+        for (Entity entity : event.getChunk().getEntities()) {
+            if (entity instanceof LivingEntity livingEntity && MobManager.getInstance()
+                                                                         .isMobManaged(entity.getUniqueId())
+            ) {
```
Comment (J3fftw1) [inline src/main/java/dev/sefiraat/netheopoiesis/listeners/ManagedMobListener.java:41]:
```suggestion
            if (entity instanceof LivingEntity livingEntity
             && MobManager.getInstance().isMobManaged(entity.getUniqueId())
            ) {
```

**Review by J3fftw1 — CHANGES_REQUESTED** [review]
quickly went over it

File: src/main/java/dev/sefiraat/netheopoiesis/api/mobs/MobCap.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+            if (entity != null) {
+                entity.remove();
+            }
+        }
+        mobs.clear();
+    }
+
+    public int getPerPlayerAmount() {
+        return perPlayerAmount;
+    }
```
Comment (J3fftw1) [inline src/main/java/dev/sefiraat/netheopoiesis/api/mobs/MobCap.java:74]: this method doesnt make sense at all
get per play amount?

File: src/main/java/dev/sefiraat/netheopoiesis/api/mobs/RandomSpawn.java [hunk 4]
```diff
@@ -138,14 +142,13 @@ public boolean trySpawn(@Nonnull Location location, int purificationLevel) {
...
 
-                LivingEntity livingEntity = (LivingEntity) location.getWorld().spawnEntity(
-                    location,
-                    this.type,
-                    this.randomize
-                );
-                livingEntity.setRemoveWhenFarAway(true);
-                livingEntity.setNoDamageTicks(20);
-                return true;
+                LivingEntity livingEntity = MobManager.getInstance().spawnMob(mobCapType, type, location, randomize);
```
Comment (J3fftw1) [inline src/main/java/dev/sefiraat/netheopoiesis/api/mobs/RandomSpawn.java:145]: no final?

**Review by J3fftw1 — CHANGES_REQUESTED** [review]

**Review by J3fftw1 — APPROVED** [review]
LGTM

**Pull request merged**